components/org.wso2.carbon.identity.auth.service/src/main/java/org/wso2/carbon/identity/auth/service/handler/AuthenticationHandler.java

```diff
@@ -101,10 +110,127 @@ protected void postAuthenticate(MessageContext messageContext, AuthenticationRes
+                            if (log.isDebugEnabled()) {
+                                log.debug("Setting the Associated user: " + associatedUser.toFullQualifiedUsername()
+                                        + ", sent with the header: "  + ASSOCIATED_USER_ID_HEADER + ", in the carbon " +
+                                        "context since it is a valid association of the user: "
+                                        + user.toFullQualifiedUsername());
+                            }
+                            setUserInCarbonContext(MultitenantUtils.getTenantAwareUsername(
+                                    associatedUser.toFullQualifiedUsername()));
+                            return;
+                        }
+                    }
+                    log.error("Associated user: " + associatedUser.toFullQualifiedUsername() + ", sent with the " +
+                            "header: "  + ASSOCIATED_USER_ID_HEADER + ", does not have a valid association with the " +
+                            "user: " + user.toFullQualifiedUsername());
+                    setFailedAuthentication(authenticationResult);
+                } catch (UserAccountAssociationException e) {
+                    log.error("Error while getting account associations of the user: "
+                            + user.toFullQualifiedUsername(), e);
+                    setFailedAuthentication(authenticationResult);
+                }
+            } else {
+                log.error("Unable to get the UserAccountConnector service");
+                setFailedAuthentication(authenticationResult);
+            }
+        } else {
+            log.error("Cannot switch to an Associated user: " + associatedUser.toFullQualifiedUsername() + ", " +
+                    "in a different tenant domain to the authenticated user: " + user.toFullQualifiedUsername());
+            setFailedAuthentication(authenticationResult);
+        }
+    }
+
+    private void setFailedAuthentication(AuthenticationResult authenticationResult) {
+
+        authenticationResult.setAuthenticationStatus(AuthenticationStatus.FAILED);
+    }
+
+    private boolean isUsersInSameTenant(User user, User associatedUser) {
+
+        return associatedUser.getTenantDomain().equals(user.getTenantDomain());
+    }
+
+    private boolean isSameUser(User user, UserAccountAssociationDTO userAccountAssociationDTO) {
+
+        User associatedUser = new User();
+        associatedUser.setTenantDomain(userAccountAssociationDTO.getTenantDomain());
+        associatedUser.setUserStoreDomain(userAccountAssociationDTO.getDomain());
+        associatedUser.setUserName(userAccountAssociationDTO.getUsername());
+        return isSameUser(user, associatedUser);
+    }
+
+    private boolean isSameUser(User firstUser, User secondUser) {
+
+        return firstUser.toFullQualifiedUsername().equals(secondUser.toFullQualifiedUsername());
+    }
+
+    private User getUser(String fullyQualifiedUserName) {
```

Comment:
Can use User.getUserFromUserName method.

Reply:
Fixed in b24e7a46ca66d94e33e0afd03cd4f787016d7325